### ISSUE TITLE
Composite: Step Outputs are not available when in nested composite actions

### ISSUE THREAD

**Issue by antdking** [actions/runner#2009]
**Describe the bug**
When a workflow uses a composite action, the post steps of actions within the Composite have their inputs evaluated with the Steps context.

However when the same Composite Action is nested in another Composite Action, the Steps context is lost for the Post steps.

Our usecase for wanting this is to cache dependencies effectively without polluting our Workflows.

**To Reproduce**
workflow uses a composite that uses a composite.

Reproduced here: https://github.com/antdking/composite-post-step-outputs-demo, specifically [this](https://github.com/antdking/composite-post-step-outputs-demo/blob/main/.github/workflows/demo-nested-composites-break.yml) workflow
This run includes debug output: https://github.com/antdking/composite-post-step-outputs-demo/actions/runs/2710027823

**Expected behavior**
Able to use Post steps for Composite Actions that are used by other Composite Actions.

## Runner Version and Platform
Hosted runner, affects all Platforms

## What's not working?
Please include error messages and screenshots.

When executing the Post Run for the wrapping composite:
```
 ##[debug]Evaluating condition for step: 'Post wrapperComposite'
##[debug]Evaluating: always()
##[debug]Evaluating always:
##[debug]=> true
##[debug]Result: true
##[debug]Starting: Post wrapperComposite
##[debug]Loading inputs
##[debug]Loading env
Post job cleanup.
##[debug]Evaluating condition for step: 'run'
##[debug]Evaluating: always()
##[debug]Evaluating always:
##[debug]=> true
##[debug]Result: true
##[debug]Starting: run
##[debug]Loading inputs
##[debug]Evaluating: steps.setter.outputs.some-val
##[debug]Evaluating Index:
##[debug]..Evaluating Index:
##[debug]....Evaluating Index:
##[debug]......Evaluating steps:
##[debug]......=> Object
##[debug]......Evaluating String:
##[debug]......=> 'setter'
##[debug]....=> null
##[debug]..=> null
##[debug]=> null
##[debug]Result: null
##[debug]Loading env
Post job cleanup.
value received: 
Error: value: '' is not 'foo'
##[debug]Node Action run completed with exit code 1
##[debug]Finished: run
##[debug]Finishing: Post wrapperComposite
```

the `setter` Step seems to be missing when evaluating the post steps.

## Runner and Worker's Diagnostic Logs

n/a

**Comment by Sytten** [comment]
I am also experiencing this issue.

Also referenced by https://github.community/t/inputs-passed-to-action-are-blank-in-post-script/262930
Minimal repro: https://github.com/mxxk/gh-actions-issue-post-inputs-missing

**Comment by antdking** [comment]
> This issue is only affecting composite actions called like `uses: ./path/to/action`, in this case the runner squashes all post steps into the outer composite action. The call flow of composite actions is hard to follow.
> 
> Caused by the ExecutionContext.Id parameter, Action.Id seems to provide better choice
> 
> https://github.com/actions/runner/blob/813af298861c7a5db62224568e0e9cb313334dab/src/Runner.Worker/ActionRunner.cs#L94
> 
> You can avoid this bug by referenceing it remotely like `uses: <myorg>/<myrepo></path/to/action>@<branch/tag/sha>`
> 
> _Does github actions have enough integration / end to end Tests?_ At least they are not publicy available

The proposed workaround won't work, as there isn't a way to pass the current git ref as a parameter (no interpolation in `uses`).

**Comment by redsun82** [comment]
I just hit this bug as well

**Comment by cognitiaclaeves** [comment]
This is a pretty nasty issue when you're trying to rely on a nested action (to get around github's lack of retry) and have no idea why it is failing.

**Comment by mkellerman-relativity** [comment]
*Bump*

**Comment by Sophie1142** [comment]
Upvoting this issue

**Comment by Yehonal** [comment]
any news here?

**Comment by haochenx** [comment]
seems like a fix is coming together with https://github.com/actions/runner/pull/1803?
is there an estimation for #1803 to be merged?

it seems that there's already a fix in https://github.com/nikola-jokic/runner/commit/69f84e99251fac57454b9aeb8c0f0daf39187380
is it possible to get it shipped first without waiting for other changes that #1803 might bring?

**Comment by Sytten** [comment]
@sethrylan Could we get some kind of response from the dev team at GitHub for this issue. It is pretty bad and the fix is a one liner...

**Comment by keradus** [comment]
For those who stumbled on using `inputs` or `outputs` for composite GitHub Actions, I found using `$GITHUB_ENV` to be a viable workaround.

You may find a working example in https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7415 .
(And also a lot of debugging in https://github.com/keradus/PHP-CS-Fixer/pull/36 .)

```yml
name: "Install composer deps"

inputs:
  flags:
    description: 'Composer flags'
    required: false

runs:
  using: "composite"
  steps:
    # Warning!
    # Use $GITUB_ENV instead of $GITHUB_OUTPUT.
    #
    # The actions/cache is, on calling it in this composite action, reading the cache,
    # and then also registering a shutdown function for whole workflow, to write the cache afterwards.
    #
    # That registered shutdown function is _NOT_ having access to inputs.* nor steps.*.outputs for
    # `path` parameter, which is evaluated on both stages (read and write), yet `key` parameters
    # are evaluated on start only and are not affected by this issue.
    #
    # Unfortunately, when the issue is faced, it does not break the build, simply raising a warning
    # and completing the build successuflly, yet without storing the cache:
    # > Warning: Input required and not supplied: path

    - name: Get Composer cache params
      shell: bash
      run: |
        echo "COMPOSER_CACHE_DIR=$(composer config cache-dir)" >> $GITHUB_ENV
        echo "COMPOSER_CACHE_PHP=$(php -r 'echo PHP_VERSION;')" >> $GITHUB_ENV

    - name: Cache Composer dependencies
      uses: actions/cache@v3
      with:
        path: ${{ env.COMPOSER_CACHE_DIR }}
        key: Composer-${{ runner.os }}-${{ env.COMPOSER_CACHE_PHP }}-${{ hashFiles('**/composer.json') }}-${{ hashFiles('**/composer.lock') }}
        restore-keys: |
          Composer-${{ runner.os }}-${{ env.COMPOSER_CACHE_PHP }}-${{ hashFiles('**/composer.json') }}-
          Composer-${{ runner.os }}-${{ env.COMPOSER_CACHE_PHP }}-
          Composer-${{ runner.os }}-

    - name: Install dependencies
      uses: nick-invision/retry@v2
      with:
        timeout_minutes: 5
        max_attempts: 5
        retry_wait_seconds: 30
        # `--no-scripts` to avoid side-effects (e.g. installing dev-tools for all jobs on CI level),
        # all executed scripts should be explicit and run only when needed.
        command: composer update --optimize-autoloader --no-interaction --no-progress --no-scripts ${{ inputs.flags }}

    - name: Show versions of packages
      shell: bash
      run: composer info -D

```

**Comment by devsh4** [comment]
Another viable workaround is to use remote actions (hosted in another repo) which worked for us as we have a central repo for reusable actions and workflows.

**Comment by amnuts** [comment]
Just hit the same issue when using `./action/name` style call.  Can't believe this is still an issue two years on!

**Comment by remidebette** [comment]
Hi, 
I encountered a similar issue when nesting an action in a reusable workflow (either calling it with absolute or relative path)
If duplicating the action code in the workflow, the problem goes away, but then exposing the issues of duplicated code.
I raised a github issue in parallel.

**Comment by VictorValar** [comment]
+1

**Comment by Galkon** [comment]
crazy this is 4 years old. just ran into it.. duping the logic is not preferred
